### PR TITLE
!!! FEATURE: Content Repository status

### LINKED ISSUE #4468: 9.0 Feature projection status and `cr:status` command
im currently building the 9.0 neos setup, and wondering if there is a way to tell the user the `cr:setup` has to be rerun - there is no way currently.

See discussion https://neos-project.slack.com/archives/C3MCBK6S2/p1691245346480469

### PR DESCRIPTION
**Pull request by bwaidelich**
Updates `neos/eventstore` and `neos/eventstore-doctrineadapter` packages to their first final release and adjusts code accordingly.

Furthermore this incorporates the following changes:

* Add `CheckpointStorageInterface` and its implementation `DbalCheckpointStorage` (aka `DoctrineCheckpointStorage`) from `neos/eventstore-doctrineadapter` (see https://github.com/neos/eventstore-doctrineadapter/pull/13)
* Add `CatchUp` from `neos/eventstore` (see https://github.com/neos/eventstore/pull/19)
* Adjust `ProjectionInterface`:
  * rename `setUp()` to `setup()` for consistency and remove return type
  * add `status()``
* Adjust `ContentRepository`:
  * rename `setUp()` to `setup()` for consistency and remove return type
  * add `status()` reflecting event store and projection statuses
* Introduce `DbalSchemaDiff` utility for central and faster schema diffing (#4762)
* Fix some projection schemas to prevent status from always returning `SETUP_REQUIRED`
* Add CLI command `cr:status`

Resolves: #4468
Related: #4762

### CODE REVIEW

File: Neos.ContentGraph.DoctrineDbalAdapter/src/DoctrineDbalContentGraphSchemaBuilder.php [hunk 1]
```diff
@@ -42,7 +42,7 @@ private function createNodeTable(): Table
             DbalSchemaFactory::columnForDimensionSpacePointHash('origindimensionspacepointhash')->setNotnull(false),
             DbalSchemaFactory::columnForNodeTypeName('nodetypename'),
             (new Column('properties', Type::getType(Types::TEXT)))->setNotnull(true)->setCustomSchemaOption('collation', self::DEFAULT_TEXT_COLLATION),
-            (new Column('classification', Type::getType(Types::STRING)))->setLength(20)->setNotnull(true)->setCustomSchemaOption('charset', 'binary'),
+            (new Column('classification', Type::getType(Types::BINARY)))->setLength(20)->setNotnull(true),
```
Comment (bwaidelich) [inline Neos.ContentGraph.DoctrineDbalAdapter/src/DoctrineDbalContentGraphSchemaBuilder.php:45]: This change is not directly related but a follow-up to #4730 – without that change, `setup()` will always drop and re-add the column (and the status will reflect that)

File: Neos.ContentRepository.Core/Classes/Infrastructure/DbalSchemaFactory.php [hunk 2]
```diff
@@ -48,9 +53,8 @@ public static function columnForNodeAggregateId(string $columnName): Column
      */
     public static function columnForContentStreamId(string $columnName): Column
     {
-        return (new Column($columnName, Type::getType(Types::STRING)))
-            ->setLength(36)
-            ->setCustomSchemaOption('charset', 'binary');
+        return (new Column($columnName, Type::getType(Types::BINARY)))
+            ->setLength(36);
```
Comment (bwaidelich) [inline Neos.ContentRepository.Core/Classes/Infrastructure/DbalSchemaFactory.php:57]: This change is not directly related but a follow-up to #4730 – without that change, `setup()` will always drop and re-add the column (and the status will reflect that)

File: Neos.Neos/Classes/AssetUsage/Projection/AssetUsageRepository.php [hunk 3]
```diff
@@ -65,13 +68,16 @@ private static function databaseSchema(string $tablePrefix): Table
...
 
-        return $table
+        $table
             ->addUniqueIndex(['assetid', 'originalassetid', 'contentstreamid', 'nodeaggregateid', 'origindimensionspacepointhash', 'propertyname'], 'assetperproperty')
             ->addIndex(['assetid'])
             ->addIndex(['originalassetid'])
             ->addIndex(['contentstreamid'])
             ->addIndex(['nodeaggregateid'])
-            ->addIndex(['origindimensionspacepointhash']);
+            ->addIndex(['origindimensionspacepointhash'], options: ['lengths' => [768]]);
```
Comment (bwaidelich) [inline Neos.Neos/Classes/AssetUsage/Projection/AssetUsageRepository.php:77]: This change is not directly related but a follow-up to #4088 – without that index length limitation, `setup()` will always drop and re-add the column (and the status will reflect that)

**Comment by mhsdesign** [comment]
WOOOOWWWW ❤️ Thanks!

Note for testing: Your local dev collection, and later the dev collection and base dist have to be adjusted:

```
    "neos/eventstore": "^1",
    "neos/eventstore-doctrineadapter": "^1",
```

Your change is non breaking to the database, so a setup was not required. And the Neos Ui e2e tests also work :D

----------


The cli output for happy case:

```
./flow cr:status
Event Store: OK

Projection "Neos\ContentRepository\Core\Projection\ContentStream\ContentStreamProjection": OK
Projection "Neos\ContentRepository\Core\Projection\Workspace\WorkspaceProjection": OK
Projection "Neos\ContentRepository\Core\Projection\NodeHiddenState\NodeHiddenStateProjection": OK
Projection "Neos\Neos\FrontendRouting\Projection\DocumentUriPathProjection": OK
Projection "Neos\Neos\PendingChangesProjection\ChangeProjection": OK
Projection "Neos\ContentRepository\Core\Projection\ContentGraph\ContentGraphProjection": OK
Projection "Neos\Neos\AssetUsage\Projection\AssetUsageProjection": OK
```


Possibly error output:
```
flow cr:status
Event Store: Setup required!
  The following SQL statements are required: ALTER TABLE cr_default_events
  CHANGE sequencenumber sequencenumber BIGINT UNSIGNED AUTO_INCREMENT NOT NULL,
  CHANGE stream stream VARCHAR(100) CHARACTER SET ascii NOT NULL, CHANGE type
  type VARCHAR(100) CHARACTER SET ascii NOT NULL, CHANGE metadata metadata
  LONGTEXT DEFAULT NULL COMMENT '(DC2Type:json)', CHANGE id id CHAR(36)
  CHARACTER SET ascii NOT NULL, CHANGE correlationid correlationid VARCHAR(40)
  CHARACTER SET ascii DEFAULT NULL, CHANGE causationid causationid VARCHAR(40)
  CHARACTER SET ascii DEFAULT NULL
  DROP INDEX id_uniq ON cr_default_events
  CREATE UNIQUE INDEX UNIQ_E2A693CBF396750 ON cr_default_events (id)
  DROP INDEX stream_version_uniq ON cr_default_events
  CREATE UNIQUE INDEX UNIQ_E2A693CF0E9BE1CBF1CD3C3 ON cr_default_events
  (stream, version)

Projection "Neos\ContentRepository\Core\Projection\ContentStream\ContentStreamProjection": OK
Projection "Neos\ContentRepository\Core\Projection\Workspace\WorkspaceProjection": Setup required!
  The following SQL statement is required: ALTER TABLE cr_default_p_workspace
  ADD isFoo TINYINT(1) NOT NULL, CHANGE baseworkspacename
  baseworkspacenameRenamed VARCHAR(255) DEFAULT NULL COLLATE
  `utf8mb4_unicode_520_ci`
Projection "Neos\ContentRepository\Core\Projection\NodeHiddenState\NodeHiddenStateProjection": OK
Projection "Neos\Neos\FrontendRouting\Projection\DocumentUriPathProjection": Setup required!
  The following SQL statement is required: ALTER TABLE
  cr_default_p_neos_documenturipath_uri CHANGE sitenodename sitenodename
  VARCHAR(500) DEFAULT '' NOT NULL COLLATE `utf8mb4_unicode_520_ci`
Projection "Neos\Neos\PendingChangesProjection\ChangeProjection": OK
Projection "Neos\ContentRepository\Core\Projection\ContentGraph\ContentGraphProjection": OK
Projection "Neos\Neos\AssetUsage\Projection\AssetUsageProjection": OK
```

This is super useful ❤️ ... and well a bit bloated as well ^^

Im in no position to propose any changes for this beautiful pr, but maybe ill do a followup so its easier to better differentiate between multiple sql lines (where one starts and the other ends) and possibly add a verbose option with 3 values:

In the normal verbosity for example only a super simple human readable output could be logged a la `Tabel cr_default_p_neos_documenturipath_uri  must be updated.`:
https://github.com/neos/neos-development-collection/blob/4473f166371c0523e04c7341621d8083d533c965/Neos.ContentRepository.Core/Classes/Projection/ProjectionStatus.php#L30-L47
And in the next level we dump the full sql lines except we trim them if they get to long or are too many.
On the maximal verbosity it will be logged like in your case currently.

And i do mean it. This can wait its more important to get this pr merged than philosophising over the verbosity of the cli output ^^

**Comment by bwaidelich** [comment]
Thanks for the nice feedback!
I definitely want to add a `--verbose` and `--quiet` flag to the CLI command!

**Comment by bwaidelich** [comment]
I split up this PR into #4845, #4846 and #4847

**Comment by bwaidelich** [comment]
> Note for testing: Your local dev collection, and later the dev collection and base dist have to be adjusted:

@mhsdesign this PR does update the [dev collection composer.json](https://github.com/neos/neos-base-distribution/blob/8.3/composer.json)!?

**Comment by mhsdesign** [comment]
Haha now i got myself confused ^^ i meant the dev distribution :D But it seems that we correctly move the dependency declaration out of there in the meantime so its all fine i guess.

**Pull request closed without merging**